src/mpid/ch4/src/ch4_types.h

```diff
@@ -68,6 +68,7 @@ typedef struct MPIDIG_hdr_t {
     MPIR_Context_id_t context_id;
     int error_bits;
     uint8_t flags;
+    MPIR_Request *sreq_ptr;
```

Comment:
Well, I guess that blows past 16-byte. We should still try ensure 16-byte alignment for performance. The issue is we embed this header else where. We should re-evaluate those cases to ensure we don't waste space unnecessarily (we can leave that for later).

PS: ideally, we would like some kind of union or simply use cast at `target_msg_cb` so that we only pay the cost of extra 16-byte when we have to (ssend, and maybe pipeline send). But that can be left as an optimization, leave a comment should suffice.

Reply:
> Well, I guess that blows past 16-byte. We should still try ensure 16-byte alignment for performance. The issue is we embed this header else where. We should re-evaluate those cases to ensure we don't waste space unnecessarily (we can leave that for later).

Yes, we should reevaluate those. Specifically, I think it makes sense to embed a "match parts" struct like how ch3 did. I think the the assumption was that `MPIDIG_hdr_t` contained only match parts, but that would change with this PR.

Reply:
I'll add a commit to remove the duplicate `sreq_ptr` from the long_req pkt type.

Reply:
Another thing we could do is send `MPI_Request` handles instead of pointer values (8 -> 4 bytes on 64-bit platforms). That is the way ch3 did it.

Reply:
> 
> 
> Another thing we could do is send `MPI_Request` handles instead of pointer values (8 -> 4 bytes on 64-bit platforms). That is the way ch3 did it.

That is a good idea. But only if we can manage to get it inside 16 bytes. Otherwise, the next padding is 32 bytes. I think the `sreq` is already optional for some protocol, and we probably will have more optional fields anyway, e.g. pipeline and rdma, so maybe it is not worthwhile to squeeze it inside the 16-byte.

>  I think it makes sense to embed a "match parts" struct like how ch3 did. I think the the assumption was that MPIDIG_hdr_t contained only match parts, but that would change with this PR.

I think that is a good idea. Having specific name for specific purpose can make the code read better.

Reply:
Let not jump into optimization too early. My observations are:
1. Only the data transmitting message need `MPIDIG_hdr_t`
2. In the main branch, only EAGER SEND has a CH4 message header that can fit in a cache line exactly. But that is assuming SSEND is handled by a separate message type which is greater than cache line size because of all the field needed at SEND itself plus the `sreq_ptr`.
3. Transmitting the handle of sreq rather than the ptr would result in a reverse handle-to-ptr translation when we need the sreq_ptr again (for example in of SSEND_ACK_TARGET_MSG_CB). This could be another source of cache miss + extra work. So I am not sure if this is a good way to save some space.
4. Combining the sync into send message would introduce a flag on top of all these. Unless we started the limit how many bits src_rank, tag can take plus transmitting request handle.

Reply:
To clarify, my comment is not about fitting into cache line. It is simply about alignment (the cacheline is typically 64bytes). If the header doesn't have 16-byte alignment, then down the path there will be branches and extra handling to pad the alignment for correctness. We should try to avoid that in the common case at least. On the other hand, if we simply force the alignment, we may incur wasted header space especially if the structure later will be embedded into other structures.

Reply:
>     1. Only the data transmitting message need `MPIDIG_hdr_t`
> 
Can we limit it further to only pt2pt sends? RMA doesn't need tag nor error bits, right?

Reply:
It already is, right? The only header type that embeds `MPIDIG_hdr_t` is `MPIDIG_send_long_req_msg_t` at the moment.

Reply:
> It already is, right? The only header type that embeds `MPIDIG_hdr_t` is `MPIDIG_send_long_req_msg_t` at the moment.

I hope so. I just want to make sure with @yfguo that we are on the same page.

Reply:
Just some more about the fields: `MPIR_Context_id_t` is `uint16_t` by default. `error_bits` could be reduced to `uint8_t`, or embedded into the tag. It is only encoding 2 possible error values at the moment.

Reply:
Yes. Right now `MPIDIG_hdr_t` === send_msg_t. We are going to need some common fields between PT2PT and RMA in order to handle the protocols. What I did is putting those commonly shared fields into `MPIDIG_hdr_t` and send will get a `MPIDIG_send_msg_t` like this:
```
typedef struct {
    int src_rank;
    uint8_t flags;
    size_t data_sz;
    MPIR_Request *sreq_ptr;
} MPIDIG_hdr_t;

struct MPIDIG_send_msg_t {
    MPIDIG_hdr_t hdr;
    MPIR_Context_id context id;
    int tag;
    int errflag;
};

struct MPIDIG_put_msg_t {
    MPIDIG_hdr_t hdr;
    // other put specific fields
};
```

Reply:
> To clarify, my comment is not about fitting into cache line. It is simply about alignment (the cacheline is typically 64bytes). If the header doesn't have 16-byte alignment, then down the path there will be branches and extra handling to pad the alignment for correctness. We should try to avoid that in the common case at least. On the other hand, if we simply force the alignment, we may incur wasted header space especially if the structure later will be embedded into other structures.

Oh, I miss-read as 16-words. I think there is a simple solution to that: just allocate the space with the size that is aligned to the MAX_ALIGNMENT. Basically this add padding spaces at the end of the entire am_hdr if needed.

Reply:
OK, what is the takeaway from this discussion? Are there additional changes to be made to this PR right now?

Reply:
I think the current PR is good to go. I will revisit the alignment issue when updating my PR

Reply:
This PR is probably good. Let me take one more review.

Reply:
> 
> 
> Yes. Right now `MPIDIG_hdr_t` === send_msg_t. We are going to need some common fields between PT2PT and RMA in order to handle the protocols. What I did is putting those commonly shared fields into `MPIDIG_hdr_t` and send will get a `MPIDIG_send_msg_t` like this:
> 
> ```
> typedef struct {
>     int src_rank;
>     uint8_t flags;
>     size_t data_sz;
>     MPIR_Request *sreq_ptr;
> } MPIDIG_hdr_t;
> 
> struct MPIDIG_send_msg_t {
>     MPIDIG_hdr_t hdr;
>     MPIR_Context_id context id;
>     int tag;
>     int errflag;
> };
> 
> struct MPIDIG_put_msg_t {
>     MPIDIG_hdr_t hdr;
>     // other put specific fields
> };
> ```

@yfguo After this PR merges, how about we have a separate PR just do your proposed header change and maybe some refactoring on the function interface moving closer to your final goal?

Reply:
@hzhou Which one are you referring to? The refactoring on the common header or the change in allocation to add padding?